test/test_pdo.py

```diff
@@ -55,6 +55,35 @@ def test_bit_mapping(self):
         self.assertEqual(node.tpdo[0x2002].raw, 0xf)
         self.assertEqual(node.pdo[0x1600][0x2002].raw, 0xf)
 
+    def test_bit_offsets(self):
+        node = canopen.Node(1, EDS_PATH)
+        pdo = node.pdo.tx[1]
+        pdo.add_variable('UNSIGNED8 value', length=4)  # byte-aligned, partial byte length
+        pdo.add_variable('INTEGER8 value')  # non-byte-aligned, one whole byte length
+        pdo.add_variable('UNSIGNED32 value', length=24)  # non-aligned, partial last byte
+        pdo.add_variable('UNSIGNED16 value', length=12)  # non-aligned, whole last byte
+        pdo.add_variable('INTEGER16 value', length=3)  # byte-aligned, partial byte length
+        pdo.add_variable('INTEGER32 value', length=13)  # non-aligned, whole last byte
+
+        # Write some values
+        pdo['UNSIGNED8 value'].raw = 3
+        pdo['INTEGER8 value'].raw = -2
+        pdo['UNSIGNED32 value'].raw = 0x987654
+        pdo['UNSIGNED16 value'].raw = 0x321
+        pdo['INTEGER16 value'].raw = -1
+        pdo['INTEGER32 value'].raw = -1071
+
+        # Check expected data
+        self.assertEqual(pdo.data, b'\xe3\x4f\x65\x87\x19\x32\x8f\xde')
```

Comment:
Has this encoding been independently verified?

It would be nice with a breakdown of the encoding in the comments here. It took me some time to recognize that the whole PDO is little endian, not just field by field.

Reply:
I calculated it with the help of a spreadsheet:

Offset | Length | Value | Value Hex | Unsigned | High DWORD | Value Hex Shifted | Low DWORD
-- | -- | -- | -- | -- | -- | -- | --
0 | 4 | 3 | 3 | 3 | 00000000 | 0000000000000003 | 00000003
4 | 8 | -2 | FE | 254 | 00000000 | 0000000000000FE0 | 00000FE0
12 | 24 | 9991764 | 987654 | 9991764 | 00000009 | 0000000987654000 | 87654000
36 | 12 | 801 | 321 | 801 | 00003210 | 0000321000000000 | 00000000
48 | 3 | -1 | 7 | 7 | 00070000 | 0007000000000000 | 00000000
51 | 13 | -1071 | 1BD1 | 7121 | DE880000 | DE88000000000000 | 00000000
64 |   |   |   |   |   |   |  

The final shifted values are then easily bit-ORed (manually) and the complete sequence reversed by bytes.

Reply:
Oddly enough, the canopen standard doesn't really seem to mention anything about non 8-bit aliged PDO encoding. I find much chatter on Google that it formally does not support this. Does this imply that there is a defacto standard for it?

(We're using a variant of canfestival (https://github.com/Laerdal/canfestival-laerdal) which does support packing)

What I'd like to double-check is the order of objects. E.g. say the PDO shall contain two 16-bit objects A and B in that order. `A = 0x1234` and `B = 0x5678`. This can either be encoded as b`\x34\x12\x78\x56` or b`\x78\x56\x34\x12`. The example above is the latter format if I'm not mistaking, but the graphics for PDO packing in the canopen standard could be interpreted to the former format. -- But I'm not certain, and I have to verify with actual HW to see what's right.

Reply:
Will have to check and verify with the standard to be sure. From memory, I recall that the device EDS specifies a granularity which sets the lower bound on mapping alignment, e.g. word, byte, nibble or bit. I don't see why it shouldn't be used with 1-bit granularity, which I've seen commonly in other CAN protocols (like J1939, NMEA2000). Note that the existing test is left untouched and its assumptions still hold. It does contain nibble mappings already.

Regarding packing order, I think it's consistent with our API and I implemented on that basis. Basically, to take a side in the Holy Endianness Wars, little-endian is the only sensible way for a good reason. The confusion stems from humans being used to write numbers (hex, decimal, whatever) in big-endian notation, which is the common oddball and makes everything look complicated when talking about little-endian format. As a reminder: little-endian means lowest address holds the least significant digit. 

Our API, and AIUI also the CANopen mapping model, starts from bit zero. Each mapping adds some bits and the running summation is the offset / address of the following mapping. So a boolean needs one byte (the first, which would be LSB in a bigger type) and occupies the lsb therein. Now we are at offset 1 and continue adding e.g. an UNSIGNED8 value, bit by bit. Starting from its lsb=0 at offset 1 up to its msb=7 at offset 8, which ends up as first bit (lsb) of the second PDO byte. Rinse and repeat. 

Those semantics are essentially the same as regarding the whole PDO as a 64-bit integer and mapping a variable by shifting its (numeric) value left (toward higher significance) as much as the cumulated bit offset indicates.

It really makes sense if you imagine every number written in binary and putting down the digits left-to-right, as you would note down a sequence of bytes. (RTL vs. LTR is really interchangeable, it only should match between writing direction in your head and number notation wrt. increasing significance.)

Given this explanation I have a hard time imagining any other logical way to pack up the PDO mappings. What was the other possible systematic again?

Reply:
*Side note to put a nail in big-endian's coffin: Little-endian not only has a linear, non-inverted relation between offset and significance (both monotonically increasing). Which is mostly cosmetic. The main advantage is that each sequentially processed bit keeps its significance over time. You always know the first bit corresponds to 2^0.*

*Whereas with big-endian you would need to know the total bit width in advance or keep counting until all bits have been received, before you know whether the first bit corresponds to 2^15 or whatever else.*

*This applies to endianness with respect to transmitted bytes just the same, of course. If that is your atomic processing width being put in a sequence.*

Reply:
Let me be neutral about it: The choice of little vs big endian is arbitrary. Guess that's why its such a popular flame war topic. Either have advantages and disadvantages. Either is used in the world around us.

Reply:
Here I have to disagree (although irrelevant to the PR). It is not arbitrary because we have TIME as a uni-directional ordering that cannot be arbitrarily reversed. So if we're talking about digits written on a piece of paper - completely arbitrary since the "address space" is freely assignable. But when doing things sequentially, it does matter if what I know at time T is well-defined, or depends on some other info that I will only know at T+x. That's the point I was making above with respect to the bit significance.

Of course, knowing the field width ahead of time (out of band) makes it just a calculation effort but otherwise equivalent. But little-endian transmission is the only sequencing that doesn't require this a priori knowledge to reach the same property.

Enter quantum computing... ;-D

Reply:
DAC and ADCs commonly use big-endian, because the biggest significance bits comes first followed by each bit of lesser and lesser influence. This is how these converters often are implemented physically.

Reply:
Very true and a good example. It definitely has the advantage of knowing the full-scale value in advance though, let alone because the voltages are strictly limited. When filling an address space of known limited size from data of the same known size, it is actually arbitrary from which end you start.